pkg/master/master.go

```diff
@@ -210,6 +214,11 @@ func setDefaults(c *Config) {
 		}
 		c.PortalNet = portalNet
 	}
+	if c.PublicServicePorts.Size == 0 {
+		// TODO: Currently no way to specify no PublicServicePorts (do we need to allow this?)
```

Comment:
High level feedback:

I really want to kill publicIPs as they exist.  I might even hedge my bet
and do a few smaller changes to limit scope on it, even though we hope to
kill it with this.

I thought more about the nesting of these vs. "a la carte" selection, sort
of what Clayton called ingress.  I think it's more powerful but somewhat
more complicated.  There's a secondary issue of "type" of service -
balanced or direct.  e.g. something like

```
Spec {
    Type ServiceType  // Balanced vs Direct
    Balanced {
        Ingress {
           ClientAffinity
           Cluster {
              IP
            }
            NodePort {}
            Public {
              IP
            }
    }
    Direct {}
}
```

Brian thinks it might be OK to make v1 completely drop PublicIPs if we can
do this.

I want to quibble about naming, once we settle on what we're doing for this
v1 moonshot.

My Monday is sort of packed, but maybe we can chat briefly and regroup.  I
may be able to "go dark" for a few days this week and try to help push this
through.

On Fri, May 8, 2015 at 3:25 PM, Tim Hockin <notifications@github.com> wrote:

> In pkg/master/master.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7679#discussion_r29981835>
> :
>
> > @@ -210,6 +214,11 @@ func setDefaults(c *Config) {
> >  		}
> >  		c.PortalNet = portalNet
> >  	}
> > +	if c.PublicServicePorts.Size == 0 {
> > +		// TODO: Currently no way to specify no PublicServicePorts (do we need to allow this?)
>
> Torn. On one hand GCE doesn't need this, so it should be disable-able. On
> the other hand, disabling this breaks the strictly-nested property of
> visibility and means that we need to check with cloud provider to determine
> if it is required. E.g. if I set visibility to LoadBalancer, AWS cloud
> provider says "must have a host port", so we fail if no public ports are
> available.
>
> After writing it out, I am OK to require it for now with a TODO that maybe
> we can support not using this on clouds that don't need it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7679/files#r29981835>
> .
>